src/devices/src/virtio/queue.rs

```diff
@@ -393,6 +398,56 @@ impl Queue {
         let addr = self.avail_ring.unchecked_add(2);
         Wrapping(mem.read_obj::<u16>(addr).unwrap())
     }
+
+    #[inline(always)]
+    pub fn used_event(&self, mem: &GuestMemoryMmap) -> Wrapping<u16> {
+        // We need to find the `used_event` field from the avail ring.
+        let used_event_addr = 4 + 2 * self.actual_size();
+
+        // Complete all the writes in add_used() before reading the event.
+        fence(Ordering::SeqCst);
+
+        let event: u16 = mem
+            .read_obj::<u16>(self.avail_ring.unchecked_add(u64::from(used_event_addr)))
+            .unwrap();
+
+        Wrapping(event)
+    }
+
+    pub fn set_avail_event(&mut self, mem: &GuestMemoryMmap) {
+        let avail_event_addr = 4 + 8 * self.actual_size();
+
+        mem.write_obj(
+            self.avail_idx(mem).0,
+            self.used_ring.unchecked_add(u64::from(avail_event_addr)),
+        )
+        .unwrap();
+
+        fence(Ordering::SeqCst);
+    }
+
+    /// Check if we need to kick the guest.
+    /// This is similar to the `vring_need_event()` method implemented by the Linux kernel.
+    pub fn needs_notification(
+        &self,
+        mem: &GuestMemoryMmap,
+        has_notification_suppression: bool,
+    ) -> bool {
+        // If the device doesn't use notification suppression, always return true
+        if !has_notification_suppression {
```

Comment:
Can we promote `has_notification_suppression` to member of Queue, or otherwise perform this check at caller level ?

Reply:
Done.